File: py-sdk/src/mcp/protocol/completion.py
"""Completion (§19).

Argument autocompletion: a best-effort, advisory facility a server offers so a client can
suggest ranked candidate values for a prompt argument or a resource-template variable. A
single ``completion/complete`` request carries the closed ``ref`` union + the partial
``argument`` (+ optional sibling ``context``); the server returns a ranked, ≤100-item list.
Gated by the ``completions`` capability.

The client-side keystroke debouncer (a UI convenience) is deferred — Python's timing model
differs from the JS ``setTimeout`` version; the protocol substance is here.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

from mcp.jsonrpc.payload import RESULT_TYPE_COMPLETE
from mcp.protocol.capability_negotiation import SERVER_METHOD_CAPABILITY, may_client_invoke, server_declares
from mcp.protocol.errors import INVALID_PARAMS_CODE

COMPLETION_COMPLETE_METHOD = "completion/complete"

#: No `completions` capability → -32601. (§19.1, R-19.1-d)
COMPLETION_METHOD_NOT_FOUND_CODE = -32601
#: Invalid params / unknown ref / unknown argument → -32602. (§19.5, R-19.5-r/-s)
COMPLETION_INVALID_PARAMS_CODE = INVALID_PARAMS_CODE
#: Internal failure computing completions → -32603. (§19.5, R-19.5-t)
COMPLETION_INTERNAL_ERROR_CODE = -32603
#: Maximum items the completion.values array may carry. (§19.4, R-19.4-c)
MAX_COMPLETION_VALUES = 100

PROMPT_REFERENCE_TYPE = "ref/prompt"
RESOURCE_TEMPLATE_REFERENCE_TYPE = "ref/resource"


# ─── Capability (§19.1) ───────────────────────────────────────────────────────

def is_valid_completions_capability(value: object) -> bool:
  """Return ``True`` for a valid ``completions`` capability — an (open) object; ``{}`` is the
  RECOMMENDED baseline. (§19.1)
  """
  return isinstance(value, dict)


def build_completions_capability() -> dict:
  """Build the RECOMMENDED baseline ``completions`` capability — ``{}``. (§19.1, R-19.1-b)"""
  return {}


def server_declares_completions(server_caps: dict) -> bool:
  """Return ``True`` when the server declares the ``completions`` capability. (§19.1)"""
  return server_declares(server_caps, "completions")


def may_call_completion(server_caps: dict) -> bool:
  """Return ``True`` when a client MAY send ``completion/complete``. (§19.1, R-19.1-c)"""
  return may_client_invoke(COMPLETION_COMPLETE_METHOD, server_caps)


def completion_gated_by_completions() -> bool:
  """Self-check: the shared gate binds ``completion/complete`` to ``completions``. (§19.1)"""
  return SERVER_METHOD_CAPABILITY.get(COMPLETION_COMPLETE_METHOD) == "completions"


# ─── Reference union (§19.3) ──────────────────────────────────────────────────

def is_valid_prompt_reference(value: object) -> bool:
  """Return ``True`` for a ``PromptReference``: ``type == "ref/prompt"`` + string ``name``. (R-19.3-a/-b)"""
  if not isinstance(value, dict) or value.get("type") != PROMPT_REFERENCE_TYPE:
    return False
  if not isinstance(value.get("name"), str):
    return False
  return "title" not in value or isinstance(value["title"], str)


def is_valid_resource_template_reference(value: object) -> bool:
  """Return ``True`` for a ``ResourceTemplateReference``: ``type == "ref/resource"`` + string
  ``uri``. (R-19.3-c/-d)
  """
  return isinstance(value, dict) and value.get("type") == RESOURCE_TEMPLATE_REFERENCE_TYPE and isinstance(value.get("uri"), str)


def is_valid_completion_reference(value: object) -> bool:
  """Return ``True`` for a member of the closed ``ref`` union; any other ``type`` is rejected.
  (§19.3, R-19.2-c–R-19.2-e, R-19.3-f)
  """
  return is_valid_prompt_reference(value) or is_valid_resource_template_reference(value)


def is_prompt_reference(ref: dict) -> bool:
  """Return ``True`` when ``ref`` is a prompt reference. (R-19.2-d)"""
  return ref.get("type") == PROMPT_REFERENCE_TYPE


def is_resource_template_reference(ref: dict) -> bool:
  """Return ``True`` when ``ref`` is a resource-template reference. (R-19.2-d)"""
  return ref.get("type") == RESOURCE_TEMPLATE_REFERENCE_TYPE


# ─── CompleteRequestParams (§19.2) ────────────────────────────────────────────

def is_valid_completion_argument(value: object) -> bool:
  """Return ``True`` for a valid ``argument``: string ``name`` + string ``value`` (MAY be ``""``).
  (R-19.2-g/-h/-i)
  """
  return isinstance(value, dict) and isinstance(value.get("name"), str) and isinstance(value.get("value"), str)


def is_valid_completion_context(value: object) -> bool:
  """Return ``True`` for a valid ``context``: OPTIONAL ``arguments`` map of string→string. (R-19.2-j)"""
  if not isinstance(value, dict):
    return False
  if "arguments" in value:
    args = value["arguments"]
    if not isinstance(args, dict) or not all(isinstance(v, str) for v in args.values()):
      return False
  return True


def is_valid_complete_request_params(value: object) -> bool:
  """Return ``True`` for valid ``completion/complete`` params: closed ``ref`` + ``argument`` +
  optional ``context``/``_meta``. (§19.2)
  """
  if not isinstance(value, dict):
    return False
  if not is_valid_completion_reference(value.get("ref")):
    return False
  if not is_valid_completion_argument(value.get("argument")):
    return False
  if "context" in value and not is_valid_completion_context(value["context"]):
    return False
  return "_meta" not in value or isinstance(value["_meta"], dict)


def build_complete_request_params(ref: dict, argument: dict, *, context: dict | None = None, meta: dict | None = None) -> dict:
  """Build ``completion/complete`` params; ``context``/``_meta`` only when supplied. (§19.2)

  :raises ValueError: when ``context.arguments`` includes the argument being completed. (R-19.2-k)
  """
  if context is not None and isinstance(context.get("arguments"), dict) and argument["name"] in context["arguments"]:
    raise ValueError(f'context.arguments MUST NOT include the argument being completed ("{argument["name"]}") (R-19.2-k)')
  params: dict = {"ref": ref, "argument": argument}
  if context is not None:
    params["context"] = context
  if meta is not None:
    params["_meta"] = meta
  return params


# ─── CompleteResult (§19.4) ───────────────────────────────────────────────────

def _is_number(value: object) -> bool:
  return isinstance(value, (int, float)) and not isinstance(value, bool)


def is_valid_completion_object(value: object) -> bool:
  """Return ``True`` for a valid ``completion`` object: ``values`` list of ≤100 strings;
  OPTIONAL ``total`` (number), ``hasMore`` (bool). (§19.4)
  """
  if not isinstance(value, dict):
    return False
  values = value.get("values")
  if not isinstance(values, list) or len(values) > MAX_COMPLETION_VALUES or not all(isinstance(v, str) for v in values):
    return False
  if "total" in value and not _is_number(value["total"]):
    return False
  return "hasMore" not in value or isinstance(value["hasMore"], bool)


def is_valid_complete_result(value: object) -> bool:
  """Return ``True`` for a well-formed ``CompleteResult``: ``resultType`` + ``completion``
  object; OPTIONAL ``_meta``. (§19.4)
  """
  if not isinstance(value, dict):
    return False
  if resolve_complete_result_type(value) != "complete":
    return False
  if not is_valid_completion_object(value.get("completion")):
    return False
  return "_meta" not in value or isinstance(value["_meta"], dict)


def resolve_complete_result_type(result: dict) -> str:
  """Resolve a result's ``resultType``, treating absent as ``"complete"``. (R-19.4-l)"""
  raw = result.get("resultType")
  return RESULT_TYPE_COMPLETE if raw is None else str(raw)


def resolve_has_more(completion: dict) -> bool:
  """Resolve the ``hasMore`` truncation hint, treating absent/non-boolean as ``False``. (R-19.4-i)"""
  return completion.get("hasMore") is True


@dataclass(frozen=True)
class CompleteResultConfig:
  """Server-supplied inputs to a ``CompleteResult``."""

  values: list
  total: int | None = None
  has_more: bool | None = None
  meta: dict | None = None


def build_complete_result(config: CompleteResultConfig) -> dict:
  """Build a successful ``CompleteResult`` (``resultType: "complete"``); optional fields only
  when supplied. (§19.4)

  :raises ValueError: when more than 100 ``values`` are supplied. (R-19.4-c/-d)
  """
  if len(config.values) > MAX_COMPLETION_VALUES:
    raise ValueError(f"CompleteResult.completion.values MUST NOT exceed {MAX_COMPLETION_VALUES} items (R-19.4-c)")
  completion: dict = {"values": list(config.values)}
  if config.total is not None:
    completion["total"] = config.total
  if config.has_more is not None:
    completion["hasMore"] = config.has_more
  result: dict = {"resultType": RESULT_TYPE_COMPLETE, "completion": completion}
  if config.meta is not None:
    result["_meta"] = config.meta
  return result


def compute_completion(ranked: list, *, total: int | None = None) -> dict:
  """Reference engine: cap an already-ranked list at 100 and signal truncation. ``total``
  defaults to ``len(ranked)``; ``hasMore`` is set when matches were dropped. (§19.4, R-19.4-c–-h)
  """
  values = list(ranked[:MAX_COMPLETION_VALUES])
  true_total = total if total is not None else len(ranked)
  completion: dict = {"values": values}
  if true_total > len(values):
    completion["total"] = true_total
    completion["hasMore"] = True
  return completion


def prefix_match(seed: str, candidates: list, *, case_insensitive: bool = False) -> list[str]:
  """Return the ``candidates`` starting with ``seed`` (input order). Empty seed matches all.
  The simplest SHOULD-permitted strategy; a server MAY substitute any matcher. (§19.5, R-19.5-d)
  """
  if seed == "":
    return list(candidates)
  needle = seed.lower() if case_insensitive else seed
  return [c for c in candidates if (c.lower() if case_insensitive else c).startswith(needle)]


# ─── Error model + validation (§19.5) ─────────────────────────────────────────

def build_completion_not_supported_error() -> dict:
  """Build the -32601 error for ``completion/complete`` without the ``completions`` capability.
  (§19.1, R-19.1-d)
  """
  return {"code": COMPLETION_METHOD_NOT_FOUND_CODE, "message": f"Method not found: {COMPLETION_COMPLETE_METHOD}"}


def build_completion_invalid_params_error(detail: str) -> dict:
  """Build a -32602 error for malformed ``completion/complete`` params. (§19.5, R-19.5-s)"""
  return {"code": COMPLETION_INVALID_PARAMS_CODE, "message": f"Invalid params: {detail}"}


def build_unknown_reference_error(detail: str) -> dict:
  """Build the -32602 error for an unknown ref / unknown argument (NOT a not-found result).
  (§19.5, R-19.5-r)
  """
  return {"code": COMPLETION_INVALID_PARAMS_CODE, "message": f"Invalid params: {detail}"}


def build_completion_internal_error(detail: str | None = None) -> dict:
  """Build the -32603 error for an internal completion failure. (§19.5, R-19.5-t)"""
  return {"code": COMPLETION_INTERNAL_ERROR_CODE, "message": f"Internal error: {detail}" if detail else "Internal error"}


@dataclass(frozen=True)
class CompleteRequestValidation:
  """Outcome of :func:`validate_complete_request`."""

  ok: bool
  params: dict | None = None
  error: dict | None = None


def validate_complete_request(params: object) -> CompleteRequestValidation:
  """Validate the SHAPE of ``completion/complete`` params (§19.2/§19.3), mapping failures to
  -32602: ``ref`` REQUIRED + closed union, ``argument`` REQUIRED string name/value, and (when
  present) ``context.arguments`` keys MUST NOT include ``argument.name``. The unknown-ref /
  unknown-argument checks need the catalog — see :func:`resolve_completion_target`.
  """
  if not isinstance(params, dict):
    return CompleteRequestValidation(False, error=build_completion_invalid_params_error("params must be an object"))
  if params.get("ref") is None:
    return CompleteRequestValidation(False, error=build_completion_invalid_params_error('"ref" is required'))
  if not is_valid_completion_reference(params.get("ref")):
    return CompleteRequestValidation(False, error=build_completion_invalid_params_error('"ref" is not a valid ref/prompt or ref/resource'))
  if not is_valid_completion_argument(params.get("argument")):
    return CompleteRequestValidation(False, error=build_completion_invalid_params_error('"argument" must have string name and value'))
  if "context" in params and not is_valid_completion_context(params["context"]):
    return CompleteRequestValidation(False, error=build_completion_invalid_params_error('"context.arguments" must be a string map'))
  ctx_args = (params.get("context") or {}).get("arguments")
  if isinstance(ctx_args, dict) and params["argument"]["name"] in ctx_args:
    return CompleteRequestValidation(
      False,
      error=build_completion_invalid_params_error(
        f'context.arguments MUST NOT include the argument being completed ("{params["argument"]["name"]}")'
      ),
    )
  return CompleteRequestValidation(True, params=params)


# ─── Reference resolution against a server catalog (§19.5) ─────────────────────

class CompletionCatalog(Protocol):
  """The server's catalog used to detect an unknown ref / unknown argument. (R-19.5-r)"""

  def prompt_argument_names(self, name: str) -> list[str] | None: ...
  def resource_template_variable_names(self, uri: str) -> list[str] | None: ...


@dataclass(frozen=True)
class CompletionTargetResolution:
  """Outcome of :func:`resolve_completion_target`."""

  ok: bool
  error: dict | None = None


def resolve_completion_target(params: dict, catalog: CompletionCatalog) -> CompletionTargetResolution:
  """Resolve a validated ``ref`` + ``argument.name`` against the server's catalog (R-19.5-r):
  an unknown prompt / template, or an argument not part of the target, is rejected with
  -32602 (NOT a not-found result).
  """
  arg_name = params["argument"]["name"]
  ref = params["ref"]
  if is_prompt_reference(ref):
    names = catalog.prompt_argument_names(ref["name"])
    if names is None:
      return CompletionTargetResolution(False, error=build_unknown_reference_error(f'unknown prompt "{ref["name"]}"'))
    if arg_name not in names:
      return CompletionTargetResolution(False, error=build_unknown_reference_error(f'prompt "{ref["name"]}" has no argument "{arg_name}"'))
    return CompletionTargetResolution(True)
  variables = catalog.resource_template_variable_names(ref["uri"])
  if variables is None:
    return CompletionTargetResolution(False, error=build_unknown_reference_error(f'unknown resource template "{ref["uri"]}"'))
  if arg_name not in variables:
    return CompletionTargetResolution(False, error=build_unknown_reference_error(f'resource template "{ref["uri"]}" has no variable "{arg_name}"'))
  return CompletionTargetResolution(True)


def prompt_argument_names_of(prompt: dict) -> list[str]:
  """Return a prompt's declared argument names, for a :class:`CompletionCatalog`. (R-19.5-r)"""
  return [arg["name"] for arg in (prompt.get("arguments") or [])]


def resource_template_variable_names_of(template: dict, extract_variables) -> list[str]:
  """Return a template's URI-template variable names, using the caller-supplied extractor
  (keep the §17.4 binding as the single source of template-variable parsing). (R-19.5-r)
  """
  return list(extract_variables(template["uriTemplate"]))

File: csharp-sdk/Protocol/Completion.cs
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

using Stackific.Mcp.JsonRpc;

namespace Stackific.Mcp.Protocol;

/// <summary>
/// The reference identifying what a <c>completion/complete</c> request is completing (spec §19.3):
/// either a prompt (<c>ref/prompt</c>) or a resource template (<c>ref/resource</c>). The closed
/// union is discriminated by the <c>type</c> field.
/// </summary>
[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(PromptReference), "ref/prompt")]
[JsonDerivedType(typeof(ResourceTemplateReference), "ref/resource")]
public abstract record CompletionReference
{
  private protected CompletionReference() { }
}

/// <summary>A completion reference identifying a prompt by name (spec §19.3, <c>type: "ref/prompt"</c>).</summary>
public sealed record PromptReference : CompletionReference
{
  /// <summary>REQUIRED. The programmatic name of the prompt being completed.</summary>
  public required string Name { get; init; }

  /// <summary>OPTIONAL. A human display name (not load-bearing for matching).</summary>
  public string? Title { get; init; }
}

/// <summary>A completion reference identifying a resource template by URI (spec §19.3, <c>type: "ref/resource"</c>).</summary>
public sealed record ResourceTemplateReference : CompletionReference
{
  /// <summary>REQUIRED. The URI or URI template whose variable is being completed.</summary>
  public required string Uri { get; init; }
}

/// <summary>The single argument being completed in a <c>completion/complete</c> request (spec §19.2).</summary>
public sealed record CompletionArgument
{
  /// <summary>REQUIRED. The name of the argument being completed.</summary>
  public required string Name { get; init; }

  /// <summary>REQUIRED. The current partial value (the match seed); MAY be empty.</summary>
  public required string Value { get; init; }
}

/// <summary>Additional disambiguating context for completion (spec §19.2).</summary>
public sealed record CompletionContext
{
  /// <summary>OPTIONAL. Already-resolved sibling arguments, keyed by name (excluding the completed argument).</summary>
  public IDictionary<string, string>? Arguments { get; init; }
}

/// <summary>The parameters of a <c>completion/complete</c> request (spec §19.2).</summary>
public sealed record CompleteRequestParams
{
  /// <summary>REQUIRED. What is being completed (a prompt or resource-template reference).</summary>
  public required CompletionReference Ref { get; init; }

  /// <summary>REQUIRED. The argument being completed.</summary>
  public required CompletionArgument Argument { get; init; }

  /// <summary>OPTIONAL. Additional context to refine suggestions.</summary>
  public CompletionContext? Context { get; init; }

  /// <summary>OPTIONAL. Reserved request metadata map (§4).</summary>
  [JsonPropertyName("_meta")]
  public JsonObject? Meta { get; init; }
}

/// <summary>The ranked candidate values returned by a completion (spec §19.4).</summary>
public sealed record CompletionValues
{
  /// <summary>REQUIRED. Candidate values, ranked by descending relevance; MUST NOT exceed 100 items.</summary>
  public required IReadOnlyList<string> Values { get; init; }

  /// <summary>OPTIONAL. The total number of matches available (MAY exceed <see cref="Values"/>).</summary>
  public int? Total { get; init; }

  /// <summary>OPTIONAL (default <c>false</c>). Whether more matches exist beyond <see cref="Values"/>.</summary>
  public bool? HasMore { get; init; }
}

/// <summary>The result of a <c>completion/complete</c> request (spec §19.4).</summary>
public sealed record CompleteResult
{
  /// <summary>REQUIRED. The completion suggestions.</summary>
  public required CompletionValues Completion { get; init; }
}

/// <summary>The set of valid argument names a completion <c>ref</c> may complete (spec §19.5, R-19.5-r).</summary>
/// <remarks>
/// Resolved from the server's catalog so <see cref="Completion.ResolveCompletionTarget"/> can detect an
/// unknown ref or an argument that is not part of the referenced target. A <see cref="PromptReference"/>
/// resolves against the server's offered prompts (by name); a <see cref="ResourceTemplateReference"/>
/// against the offered resource templates (by uri/uriTemplate). A target found but carrying an empty
/// argument-name set is still "known" — only an absent target is unknown.
/// </remarks>
public interface ICompletionCatalog
{
  /// <summary>Resolves the declared argument names of a prompt, or <c>null</c> when the prompt is unknown.</summary>
  /// <param name="name">The prompt name.</param>
  /// <returns>The declared argument names, or <c>null</c> when unknown.</returns>
  IReadOnlyList<string>? PromptArgumentNames(string name);

  /// <summary>Resolves the declared variable names of a resource template, or <c>null</c> when unknown.</summary>
  /// <param name="uri">The template's URI / URI template.</param>
  /// <returns>The declared variable names, or <c>null</c> when unknown.</returns>
  IReadOnlyList<string>? ResourceTemplateVariableNames(string uri);
}

/// <summary>The discriminated outcome of <see cref="Completion.ResolveCompletionTarget"/> (spec §19.5, R-19.5-r).</summary>
/// <param name="Ok">Whether the ref and argument resolve against the catalog.</param>
/// <param name="Error">The <c>-32602</c> error when resolution fails.</param>
public readonly record struct CompletionTargetResolution(bool Ok, McpError? Error)
{
  /// <summary>A successful resolution.</summary>
  public static CompletionTargetResolution Valid { get; } = new(true, null);

  /// <summary>Builds a failed resolution carrying the <c>-32602</c> <paramref name="error"/>.</summary>
  /// <param name="error">The invalid-params error.</param>
  /// <returns>A failed resolution.</returns>
  public static CompletionTargetResolution Invalid(McpError error) => new(false, error);
}

/// <summary>
/// The §19.4–§19.5 normative completion helpers ported from the TypeScript SDK's <c>completion.ts</c>:
/// the 100-item cap with truncation signalling (<see cref="ComputeCompletion"/>, only emitting
/// <c>total</c>/<c>hasMore</c> when matches are dropped), the prefix matcher, the context
/// key-exclusion guard (R-19.2-k), and the unknown-ref / unknown-argument resolution that maps to
/// <c>-32602</c> rather than a not-found result (R-19.5-r).
/// </summary>
public static class Completion
{
  /// <summary>The maximum number of items <c>completion.values</c> may carry (§19.4, R-19.4-c).</summary>
  public const int MaxCompletionValues = 100;

  /// <summary>
  /// Caps an already-ranked candidate list at 100 and signals truncation, producing the <c>completion</c>
  /// object a server returns. Critically — matching the TS <c>computeCompletion</c> — <c>total</c> and
  /// <c>hasMore</c> are set ONLY when matches were dropped (<c>total &gt; values.Count</c>); an
  /// under-cap result leaves them ABSENT (unknown), rather than always emitting an exact <c>total</c>.
  /// (§19.4, R-19.4-c – R-19.4-h, R-19.5-g, R-19.5-h)
  /// </summary>
  /// <param name="ranked">Candidate values already ordered by descending relevance.</param>
  /// <param name="totalOverride">The true match count when <paramref name="ranked"/> is itself a pre-truncated subset; takes precedence over its length.</param>
  /// <returns>The <see cref="CompletionValues"/> with the cap applied and truncation signalled only when dropped.</returns>
  public static CompletionValues ComputeCompletion(IReadOnlyList<string> ranked, int? totalOverride = null)
  {
    ArgumentNullException.ThrowIfNull(ranked);
    var values = ranked.Count > MaxCompletionValues
      ? ranked.Take(MaxCompletionValues).ToList()
      : ranked.ToList();

    // The true match total: an explicit override (caller knows of more than it materialized) takes
    // precedence over the length of the supplied list.
    var trueTotal = totalOverride ?? ranked.Count;
    if (trueTotal > values.Count)
    {
      return new CompletionValues { Values = values, Total = trueTotal, HasMore = true };
    }

    return new CompletionValues { Values = values };
  }

  /// <summary>
  /// A reference prefix matcher: returns the <paramref name="candidates"/> whose value starts with the
  /// <paramref name="seed"/>, in input order. When the seed is empty, every candidate matches (the
  /// empty-input suggestions). Matching is case-sensitive by default. (§19.5, R-19.5-d, R-19.2-i)
  /// </summary>
  /// <param name="seed">The current partial value (<c>argument.value</c>).</param>
  /// <param name="candidates">The full candidate pool.</param>
  /// <param name="caseInsensitive">When <c>true</c>, folds case before matching.</param>
  /// <returns>The matching candidates, in input order.</returns>
  public static IReadOnlyList<string> PrefixMatch(string seed, IReadOnlyList<string> candidates, bool caseInsensitive = false)
  {
    ArgumentNullException.ThrowIfNull(seed);
    ArgumentNullException.ThrowIfNull(candidates);
    if (seed.Length == 0) return candidates.ToList();
    var comparison = caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
    return candidates.Where(c => c.StartsWith(seed, comparison)).ToList();
  }

  /// <summary>
  /// Resolves the <c>resultType</c> of a received <c>completion/complete</c> result, treating an absent
  /// (or non-string) value as <c>"complete"</c>. (§19.4, R-19.4-l)
  /// </summary>
  /// <param name="result">The raw result object received on the wire.</param>
  /// <returns>The resolved result type.</returns>
  public static string ResolveResultType(JsonObject result)
  {
    ArgumentNullException.ThrowIfNull(result);
    return result["resultType"] is JsonValue v && v.GetValueKind() == System.Text.Json.JsonValueKind.String
      ? v.GetValue<string>()
      : ResultTypes.Complete;
  }

  /// <summary>
  /// Resolves the <c>hasMore</c> truncation hint, treating an absent (or non-<c>true</c>) value as
  /// <c>false</c>. (§19.4, R-19.4-i)
  /// </summary>
  /// <param name="hasMore">The wire <c>hasMore</c> value (nullable).</param>
  /// <returns><c>true</c> only when <paramref name="hasMore"/> is explicitly <c>true</c>.</returns>
  public static bool ResolveHasMore(bool? hasMore) => hasMore == true;

  /// <summary>
  /// Enforces the R-19.2-k guard: a <c>context.arguments</c> key MUST NOT name the argument being
  /// completed. Throws <c>-32602</c> (Invalid params) when it does. A server MAY otherwise ignore
  /// <c>context</c> entirely. (§19.2, R-19.2-j, R-19.2-k)
  /// </summary>
  /// <param name="argument">The argument being completed.</param>
  /// <param name="context">The optional completion context.</param>
  /// <exception cref="McpError">A <c>-32602</c> error when the context names the completed argument.</exception>
  public static void GuardContextExcludesArgument(CompletionArgument argument, CompletionContext? context)
  {
    ArgumentNullException.ThrowIfNull(argument);
    if (context?.Arguments is { } siblings && siblings.ContainsKey(argument.Name))
    {
      throw McpError.InvalidParams(
        $"context.arguments MUST NOT include the argument being completed (\"{argument.Name}\") (R-19.2-k)");
    }
  }

  /// <summary>
  /// Resolves a validated <c>ref</c> + <c>argument.name</c> against the server's catalog, enforcing
  /// R-19.5-r: an unknown prompt / resource template, OR an <c>argument.name</c> that is not a declared
  /// argument/variable of the referenced target, MUST be rejected with <c>-32602</c> (Invalid params)
  /// — NOT a not-found result. A KNOWN argument with no registered completer is NOT an error; it simply
  /// yields empty values. (§19.5, R-19.5-r)
  /// </summary>
  /// <param name="reference">The completion reference (selected by <c>type</c>).</param>
  /// <param name="argumentName">The argument/variable name being completed.</param>
  /// <param name="catalog">The server's prompt / resource-template catalog.</param>
  /// <returns>A <see cref="CompletionTargetResolution"/>; on failure it carries the <c>-32602</c> error.</returns>
  public static CompletionTargetResolution ResolveCompletionTarget(
    CompletionReference reference,
    string argumentName,
    ICompletionCatalog catalog)
  {
    ArgumentNullException.ThrowIfNull(reference);
    ArgumentNullException.ThrowIfNull(argumentName);
    ArgumentNullException.ThrowIfNull(catalog);

    switch (reference)
    {
      case PromptReference prompt:
        {
          var names = catalog.PromptArgumentNames(prompt.Name);
          if (names is null)
          {
            return CompletionTargetResolution.Invalid(
              McpError.InvalidParams($"unknown prompt \"{prompt.Name}\"", new JsonObject { ["promptName"] = prompt.Name }));
          }

          if (!names.Contains(argumentName))
          {
            return CompletionTargetResolution.Invalid(
              McpError.InvalidParams($"prompt \"{prompt.Name}\" has no argument \"{argumentName}\""));
          }

          return CompletionTargetResolution.Valid;
        }

      case ResourceTemplateReference template:
        {
          var variables = catalog.ResourceTemplateVariableNames(template.Uri);
          if (variables is null)
          {
            return CompletionTargetResolution.Invalid(
              McpError.InvalidParams($"unknown resource template \"{template.Uri}\"", new JsonObject { ["uri"] = template.Uri }));
          }

          if (!variables.Contains(argumentName))
          {
            return CompletionTargetResolution.Invalid(
              McpError.InvalidParams($"resource template \"{template.Uri}\" has no variable \"{argumentName}\""));
          }

          return CompletionTargetResolution.Valid;
        }

      default:
        return CompletionTargetResolution.Invalid(McpError.InvalidParams("Unknown completion reference type."));
    }
  }

  /// <summary>
  /// Creates a client-side debouncer that coalesces rapid successive completion calls (for example one
  /// per keystroke) into a single in-flight <c>completion/complete</c> request: each call resets a
  /// <paramref name="waitMs"/> timer, and only the final value after a quiet period is sent. All callers
  /// awaiting during a burst resolve with that single result. (§19.5, R-19.5-n — SHOULD)
  /// </summary>
  /// <remarks>
  /// The C# counterpart of the TS <c>createCompletionDebouncer</c>. Returns a delegate that maps an
  /// argument value to a <see cref="Task{TResult}"/>; the returned debouncer is itself thread-safe and
  /// may be reused across bursts. Cancellation/disposal is the caller's concern (no timer leaks: the
  /// internal timer is one-shot and re-armed per call).
  /// </remarks>
  /// <typeparam name="T">The completion result type.</typeparam>
  /// <param name="run">Issues the actual completion request for an argument value.</param>
  /// <param name="waitMs">The quiet period (ms) before the coalesced call fires. Default 150.</param>
  /// <returns>A debounced runner mapping a value to a task that resolves with the coalesced result.</returns>
  public static Func<string, Task<T>> CreateCompletionDebouncer<T>(Func<string, Task<T>> run, int waitMs = 150)
  {
    ArgumentNullException.ThrowIfNull(run);
    return new CompletionDebouncer<T>(run, waitMs).Invoke;
  }
}

/// <summary>
/// The stateful machinery behind <see cref="Completion.CreateCompletionDebouncer{T}"/> (spec §19.5,
/// R-19.5-n): coalesces a burst of calls into one invocation of the wrapped runner with the final
/// value, resolving every awaiting caller with the single coalesced result.
/// </summary>
internal sealed class CompletionDebouncer<T>
{
  private readonly Func<string, Task<T>> _run;
  private readonly int _waitMs;
  private readonly object _gate = new();
  private readonly List<TaskCompletionSource<T>> _waiters = [];
  private CancellationTokenSource? _timer;

  public CompletionDebouncer(Func<string, Task<T>> run, int waitMs)
  {
    _run = run;
    _waitMs = waitMs;
  }

  /// <summary>
  /// Records a keystroke value, (re)arming the quiet-period timer; the returned task resolves with the
  /// coalesced result when the burst settles.
  /// </summary>
  /// <param name="value">The latest partial argument value.</param>
  /// <returns>A task resolving with the single coalesced completion result.</returns>
  public Task<T> Invoke(string value)
  {
    var waiter = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
    CancellationTokenSource cts;
    lock (_gate)
    {
      _waiters.Add(waiter);
      _timer?.Cancel();
      _timer?.Dispose();
      cts = new CancellationTokenSource();
      _timer = cts;
    }

    // Fire after the quiet window unless a later call cancels this timer first.
    _ = Task.Delay(_waitMs, cts.Token).ContinueWith(
      delayTask =>
      {
        if (delayTask.IsCanceled) return; // superseded by a later keystroke
        Fire(value);
      },
      CancellationToken.None,
      TaskContinuationOptions.ExecuteSynchronously,
      TaskScheduler.Default);

    return waiter.Task;
  }

  private void Fire(string value)
  {
    List<TaskCompletionSource<T>> batch;
    lock (_gate)
    {
      batch = [.. _waiters];
      _waiters.Clear();
      _timer?.Dispose();
      _timer = null;
    }

    if (batch.Count == 0) return;

    _run(value).ContinueWith(
      task =>
      {
        if (task.IsFaulted)
        {
          var error = task.Exception?.GetBaseException() ?? new InvalidOperationException("completion runner faulted");
          foreach (var w in batch) w.TrySetException(error);
        }
        else if (task.IsCanceled)
        {
          foreach (var w in batch) w.TrySetCanceled();
        }
        else
        {
          foreach (var w in batch) w.TrySetResult(task.Result);
        }
      },
      CancellationToken.None,
      TaskContinuationOptions.ExecuteSynchronously,
      TaskScheduler.Default);
  }
}
